pkg/controller/webhookbootstrap/controller.go

```diff
@@ -0,0 +1,488 @@
+		return true
+	}
+	cert, err := pki.DecodeX509CertificateBytes(crtData)
+	if err != nil {
+		log.Info("Re-issuing webhook certificate")
+		return true
+	}
+
+	// ensure private key is valid for certificate
+	matches, err := pki.PublicKeyMatchesCertificate(pk.Public(), cert)
+	if err != nil {
+		log.Error(err, "internal error checking certificate, re-issuing certificate")
+		return true
+	}
+	if !matches {
+		log.Info("certificate does not match private key, re-issuing")
+		return true
+	}
+
+	// validate the common name is correct
+	expectedCN := pki.CommonNameForCertificate(crt)
+	if expectedCN != cert.Subject.CommonName {
+		log.Info("certificate common name is not as expected, re-issuing")
+		return true
+	}
+
+	// validate the dns names are correct
+	expectedDNSNames := pki.DNSNamesForCertificate(crt)
+	if !util.EqualUnsorted(cert.DNSNames, expectedDNSNames) {
+		log.Info("certificate dns names are not as expected, re-issuing")
+		return true
+	}
+
+	// validate the ip addresses are correct
+	if !util.EqualUnsorted(pki.IPAddressesToString(cert.IPAddresses), crt.Spec.IPAddresses) {
+		log.Info("certificate ip addresses are not as expected, re-issuing")
+		return true
+	}
+
+	if c.certificateNeedsRenew(ctx, cert, crt) {
+		log.Info("certificate requires renewal, re-issuing")
+		return true
+	}
+
+	return false
+}
+
+func readSecretDataKey(secret *corev1.Secret, key string) []byte {
+	if secret.Data == nil {
+		return nil
+	}
+	d, ok := secret.Data[key]
+	if !ok {
+		return nil
+	}
+	return d
+}
+
+func (c *controller) generatePrivateKey(crt *cmapi.Certificate, secret *corev1.Secret) error {
+	pk, err := c.generatePrivateKeyBytes(crt)
+	if err != nil {
+		return err
+	}
+
+	return c.updateSecret(secret, pk, nil, nil)
+}
+
+func (c *controller) selfSignCertificate(crt *cmapi.Certificate, signeeKey crypto.Signer) ([]byte, error) {
+	return c.signCertificate(crt, signeeKey, signeeKey, nil)
+}
+
+func (c *controller) updateSecret(secret *corev1.Secret, pk, ca, crt []byte) error {
+	secret = secret.DeepCopy()
+	if secret.Annotations == nil {
+		secret.Annotations = make(map[string]string)
+	}
+	secret.Annotations[cmapi.AllowsInjectionFromSecretAnnotation] = "true"
+	if secret.Data == nil {
+		secret.Data = make(map[string][]byte)
+	}
+	secret.Data[corev1.TLSPrivateKeyKey] = pk
+	secret.Data[corev1.TLSCertKey] = crt
+	secret.Data[cmapi.TLSCAKey] = ca
+	_, err := c.kubeClient.CoreV1().Secrets(secret.Namespace).Update(secret)
+	return err
+}
+
+// ensureSecretsExist ensures that the webhook secrets actually exist.
+// This is to ensure that the ProcessItem function is actually called with the
+// webhook's Secret resource, so that it can be provisioned.
+func (c *controller) ensureSecretsExist(ctx context.Context) {
+	// TODO: we should be able to just not run the controller at all if these
+	//       are not set, but for now we add this hacky check.
+	if c.webhookNamespace == "" || c.webhookCASecret == "" || c.webhookServingSecret == "" {
```

Comment:
Unless I'm mistaken, we should need to do this check here since we have this is ProcessItem
```
	if c.webhookNamespace != namespace || !(c.webhookCASecret == name || c.webhookServingSecret == name) {
		return nil
	}

```

Reply:
This function isn't called by ProcessItem, so we do need to check again.